### ISSUE TITLE
bindPort{TCP,Gen,GenEx} only accepts numeric hostnames 

### ISSUE THREAD

**Issue by TravisCardwell** [fpco/streaming-commons#46]
A colleague recently discovered that he is unable to use Warp with hostnames (such as "localhost"); it only works with IP addresses.

1. `runSettings` calls `bindPortTCP` ([source](https://github.com/yesodweb/wai/blob/warp/3.2.13/warp/Network/Wai/Handler/Warp/Run.hs#L108))
2. `bindPortTCP` calls `bindPortGen` ([source](https://github.com/fpco/streaming-commons/blob/streaming-commons/0.1.18/Data/Streaming/Network.hs#L451))
3. `bindPortGen` calls `bindPortGenEx` ([source](https://github.com/fpco/streaming-commons/blob/streaming-commons/0.1.18/Data/Streaming/Network.hs#L151))
4. `bindPortGenEx` sets the `NS.AI_NUMERICHOST` hint ([source](https://github.com/fpco/streaming-commons/blob/streaming-commons/0.1.18/Data/Streaming/Network.hs#L162)), which limits hosts to numeric addresses

Is this a desired limitation?  If so, perhaps it should be documented, as none of the above functions have documentation that mentions it.  If not, would removing the `NS.AI_NUMERICHOST` hint be an acceptable fix, or would that have undesired consequences for other uses of the function?

Note that the limitation can be easily worked around by resolving hostnames before configuring the Warp settings.

For anybody who runs across this issue and searches for it, here is the error message:

```
Network.Socket.getAddrInfo (called with
preferred socket type/protocol: AddrInfo
  { addrFlags = [AI_PASSIVE,AI_NUMERICSERV,AI_NUMERICHOST]
  , addrFamily = AF_UNSPEC
  , addrSocketType = Stream
  , addrProtocol = 0
  , addrAddress = <assumed to be undefined>
  , addrCanonName = <assumed to be undefined>
  },
host name: Just "localhost",
service name: Just "8000"):
does not exist (Name or service not known)
```

**Comment by snoyberg** [comment]
That set of hints has been floating through my network code for a very long time, and I can't say I know what negative impacts could come from changing them. The earliest commit I can find with them is:

https://github.com/yesodweb/wai/commit/02c1396c86e3fceb48cbe7df58cb631c804e24d4

@kazu-yamamoto Do you have any thoughts here?

**Comment by kazu-yamamoto** [comment]
I would suggest `addrFlags = [AI_PASSIVE,AI_ADDRCONFIG]`.

- Delete `AI_NUMERICHOST` to allow hostnames
- Delete `AI_NUMERICSERV` to allow portnames
- Add `AI_ADDRCONFIG` to ensure to have suitable IP addresses

**Comment by snoyberg** [comment]
I'm fine with such a change. Travis: you want to try this out and, if it
works for you, send a PR with the update?

**Comment by TravisCardwell** [comment]
Kazu's suggestion sounds good to me.  I tested against my use cases, and it works for me.  I have submitted the changes as PR #47.

**Comment by snoyberg** [comment]
Addressed by #47, and uploaded as 0.1.19, thanks!
